File: facturacion-back/facturacion-back/src/main/java/com/cibercom/facturacion_back/service/ClienteCatalogoService.java
package com.cibercom.facturacion_back.service;

import com.cibercom.facturacion_back.model.ClienteCatalogo;
import com.cibercom.facturacion_back.repository.ClienteCatalogoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class ClienteCatalogoService {

    @Autowired
    private ClienteCatalogoRepository clienteCatalogoRepository;

    @Autowired
    private JdbcTemplate jdbcTemplate;

    private static final Logger log = LoggerFactory.getLogger(ClienteCatalogoService.class);

    public Optional<ClienteCatalogo> buscarPorRfc(String rfc) {
        if (!StringUtils.hasText(rfc)) {
            log.warn("Búsqueda cliente catálogo: RFC vacío o nulo");
            return Optional.empty();
        }
        String normalized = rfc.trim().toUpperCase();
        log.info("Consultando tabla CLIENTES por RFC: {}", normalized);
        
        // Primero intentar búsqueda exacta normalizada
        Optional<ClienteCatalogo> result = clienteCatalogoRepository.findByRfcNormalized(normalized);
        if (result.isPresent()) {
            log.info("CLIENTES RFC {} encontrado con igualdad exacta normalizada", normalized);
            return result;
        }
        
        // Si no se encuentra, intentar ignorar prefijo
        log.info("CLIENTES RFC {} no encontrado con igualdad exacta; probando ignorar prefijo", normalized);
        result = clienteCatalogoRepository.findByRfcOrPrefixed(normalized);
        if (result.isPresent()) {
            log.info("CLIENTES RFC {} encontrado con búsqueda que ignora prefijo", normalized);
            return result;
        }
        
        // Si aún no se encuentra, intentar búsqueda parcial (solo si el RFC tiene al menos 3 caracteres)
        if (normalized.length() >= 3) {
            log.info("CLIENTES RFC {} no encontrado con búsqueda exacta; probando búsqueda parcial", normalized);
            List<ClienteCatalogo> resultadosParciales = buscarPorRFCParcial(normalized, 10);
            // Buscar coincidencia exacta en los resultados parciales
            for (ClienteCatalogo c : resultadosParciales) {
                if (c.getRfc() != null && c.getRfc().trim().toUpperCase().equals(normalized)) {
                    log.info("CLIENTES RFC {} encontrado en búsqueda parcial con coincidencia exacta", normalized);
                    return Optional.of(c);
                }
            }
            // Si hay un solo resultado parcial y es muy similar, usarlo
            if (resultadosParciales.size() == 1) {
                ClienteCatalogo unico = resultadosParciales.get(0);
                String rfcUnico = unico.getRfc() != null ? unico.getRfc().trim().toUpperCase() : "";
                // Verificar si el RFC buscado está contenido en el encontrado o viceversa
                if (rfcUnico.contains(normalized) || normalized.contains(rfcUnico)) {
                    log.info("CLIENTES RFC {} encontrado en búsqueda parcial con coincidencia parcial (único resultado)", normalized);
                    return Optional.of(unico);
                }
            }
        }
        
        log.info("Resultado CLIENTES RFC {}: NO_ENCONTRADO", normalized);
        return Optional.empty();
    }

    public long contar() {
        return clienteCatalogoRepository.count();
    }

    public Map<String, Object> diagnosticoDb() {
        Map<String, Object> info = new HashMap<>();
        try {
            String user = jdbcTemplate.queryForObject("select user from dual", String.class);
            String conName = jdbcTemplate.queryForObject("select sys_context('USERENV','CON_NAME') from dual", String.class);
            String currentSchema = jdbcTemplate.queryForObject("select sys_context('USERENV','CURRENT_SCHEMA') from dual", String.class);
            info.put("user", user);
            info.put("pdb", conName);
            info.put("currentSchema", currentSchema);
        } catch (Exception e) {
            info.put("user_error", e.getMessage());
        }
        try {
            Long countNative = jdbcTemplate.queryForObject("select count(*) from CLIENTES", Long.class);
            info.put("clientesCountNative", countNative);
        } catch (Exception e) {
            info.put("clientesCountError", e.getMessage());
        }
        try {
            List<Map<String, Object>> owners = jdbcTemplate.queryForList("select owner, table_name from all_tables where table_name = 'CLIENTES'");
            info.put("owners", owners);
        } catch (Exception e) {
            info.put("owners_error", e.getMessage());
        }
        try {
            List<Map<String, Object>> synonyms = jdbcTemplate.queryForList("select owner, synonym_name, table_owner, table_name from all_synonyms where synonym_name = 'CLIENTES'");
            info.put("synonyms", synonyms);
        } catch (Exception e) {
            info.put("synonyms_error", e.getMessage());
        }
        log.info("Diagnóstico DB: {}", info);
        return info;
    }

    public ClienteCatalogo guardar(ClienteCatalogo cliente) {
        log.info("Guardando cliente en CLIENTES RFC: {}", cliente != null ? cliente.getRfc() : "null");
        return clienteCatalogoRepository.save(cliente);
    }

    public Map<String, Object> clientesMuestra(int limit) {
        Map<String, Object> out = new HashMap<>();
        int n = Math.max(limit, 1);
        try {
            Long countNative = jdbcTemplate.queryForObject("select count(*) from CLIENTES", Long.class);
            out.put("count", countNative);
        } catch (Exception e) {
            out.put("count_error", e.getMessage());
        }
        try {
            List<Map<String, Object>> rows = jdbcTemplate.query(
                    "select ID_CLIENTE, RFC, RAZON_SOCIAL from CLIENTES where ROWNUM <= ?",
                    ps -> ps.setInt(1, n),
                    (rs, rowNum) -> {
                        Map<String, Object> m = new HashMap<>();
                        m.put("idCliente", rs.getObject("ID_CLIENTE"));
                        m.put("rfc", rs.getString("RFC"));
                        m.put("razonSocial", rs.getString("RAZON_SOCIAL"));
                        return m;
                    }
            );
            out.put("rows", rows);
        } catch (Exception e) {
            out.put("rows_error", e.getMessage());
        }
        try {
            List<Map<String, Object>> owners = jdbcTemplate.queryForList("select owner, table_name from all_tables where table_name = 'CLIENTES'");
            out.put("owners", owners);
        } catch (Exception e) {
            out.put("owners_error", e.getMessage());
        }
        log.info("CLIENTES muestra: {}", out);
        return out;
    }

    /**
     * Busca clientes por RFC parcial (para autocompletado)
     * @param rfcParcial RFC parcial a buscar (mínimo 3 caracteres)
     * @param limit Límite de resultados (default 10)
     * @return Lista de clientes que coinciden con el RFC parcial
     */
    public List<ClienteCatalogo> buscarPorRFCParcial(String rfcParcial, int limit) {
        if (!StringUtils.hasText(rfcParcial) || rfcParcial.trim().length() < 3) {
            log.warn("Búsqueda parcial RFC: texto muy corto o vacío");
            return List.of();
        }
        String normalized = rfcParcial.trim().toUpperCase();
        int maxResults = Math.max(1, Math.min(limit, 50)); // Limitar entre 1 y 50
        
        try {
            List<ClienteCatalogo> resultados = jdbcTemplate.query(
                    "SELECT * FROM (SELECT * FROM CLIENTES WHERE UPPER(RFC) LIKE ? ORDER BY RFC) WHERE ROWNUM <= ?",
                    ps -> {
                        ps.setString(1, "%" + normalized + "%");
                        ps.setInt(2, maxResults);
                    },
                    (rs, rowNum) -> {
                        ClienteCatalogo c = new ClienteCatalogo();
                        c.setIdCliente(rs.getLong("ID_CLIENTE"));
                        c.setRfc(rs.getString("RFC"));
                        c.setRazonSocial(rs.getString("RAZON_SOCIAL"));
                        c.setNombre(rs.getString("NOMBRE"));
                        c.setPaterno(rs.getString("PATERNO"));
                        c.setMaterno(rs.getString("MATERNO"));
                        c.setCorreoElectronico(rs.getString("CORREO_ELECTRONICO"));
                        c.setDomicilioFiscal(rs.getString("DOMICILIO_FISCAL"));
                        c.setRegimenFiscal(rs.getString("REGIMEN_FISCAL"));
                        c.setPais(rs.getString("PAIS"));
                        c.setRegistroTributario(rs.getString("REGISTRO_TRIBUTARIO"));
                        c.setUsoCfdi(rs.getString("USO_CFDI"));
                        return c;
                    }
            );
            log.info("Búsqueda parcial RFC '{}': {} resultados", normalized, resultados.size());
            return resultados;
        } catch (Exception e) {
            log.error("Error en búsqueda parcial RFC", e);
            return List.of();
        }
    }

    /**
     * Obtiene todos los clientes
     */
    public List<ClienteCatalogo> obtenerTodos() {
        try {
            return clienteCatalogoRepository.findAll();
        } catch (Exception e) {
            log.error("Error al obtener todos los clientes", e);
            return List.of();
        }
    }

    /**
     * Obtiene clientes paginados ordenados por fecha de alta descendente (más nuevos primero)
     * @param page Número de página (0-indexed)
     * @param size Tamaño de página
     * @return Lista de clientes para la página solicitada
     */
    public List<ClienteCatalogo> obtenerTodosPaginados(int page, int size) {
        try {
            int offset = page * size;
            // Oracle paginación usando subconsulta con ROWNUM
            // Especificamos las columnas explícitamente para evitar problemas
            String sql = "SELECT ID_CLIENTE, RFC, RAZON_SOCIAL, NOMBRE, PATERNO, MATERNO, " +
                        "CORREO_ELECTRONICO, DOMICILIO_FISCAL, REGIMEN_FISCAL, PAIS, " +
                        "REGISTRO_TRIBUTARIO, USO_CFDI, FECHA_ALTA FROM (" +
                        "  SELECT a.*, ROWNUM rnum FROM (" +
                        "    SELECT ID_CLIENTE, RFC, RAZON_SOCIAL, NOMBRE, PATERNO, MATERNO, " +
                        "           CORREO_ELECTRONICO, DOMICILIO_FISCAL, REGIMEN_FISCAL, PAIS, " +
                        "           REGISTRO_TRIBUTARIO, USO_CFDI, FECHA_ALTA " +
                        "    FROM CLIENTES " +
                        "    ORDER BY FECHA_ALTA DESC NULLS LAST, ID_CLIENTE DESC" +
                        "  ) a WHERE ROWNUM <= ?" +
                        ") WHERE rnum > ?";
            
            return jdbcTemplate.query(
                    sql,
                    ps -> {
                        ps.setInt(1, offset + size);
                        ps.setInt(2, offset);
                    },
                    (rs, rowNum) -> {
                        ClienteCatalogo c = new ClienteCatalogo();
                        c.setIdCliente(rs.getLong("ID_CLIENTE"));
                        c.setRfc(rs.getString("RFC"));
                        c.setRazonSocial(rs.getString("RAZON_SOCIAL"));
                        c.setNombre(rs.getString("NOMBRE"));
                        c.setPaterno(rs.getString("PATERNO"));
                        c.setMaterno(rs.getString("MATERNO"));
                        c.setCorreoElectronico(rs.getString("CORREO_ELECTRONICO"));
                        c.setDomicilioFiscal(rs.getString("DOMICILIO_FISCAL"));
                        c.setRegimenFiscal(rs.getString("REGIMEN_FISCAL"));
                        c.setPais(rs.getString("PAIS"));
                        c.setRegistroTributario(rs.getString("REGISTRO_TRIBUTARIO"));
                        c.setUsoCfdi(rs.getString("USO_CFDI"));
                        java.sql.Timestamp fechaAlta = rs.getTimestamp("FECHA_ALTA");
                        if (fechaAlta != null) {
                            c.setFechaAlta(fechaAlta.toLocalDateTime());
                        }
                        return c;
                    }
            );
        } catch (Exception e) {
            log.error("Error al obtener clientes paginados", e);
            return List.of();
        }
    }

    /**
     * Cuenta el total de clientes
     */
    public long contarTotal() {
        try {
            return clienteCatalogoRepository.count();
        } catch (Exception e) {
            log.error("Error al contar clientes", e);
            return 0;
        }
    }

    /**
     * Obtiene un cliente por ID
     */
    public Optional<ClienteCatalogo> obtenerPorId(Long id) {
        if (id == null) {
            return Optional.empty();
        }
        try {
            return clienteCatalogoRepository.findById(id);
        } catch (Exception e) {
            log.error("Error al obtener cliente por ID: {}", id, e);
            return Optional.empty();
        }
    }

    /**
     * Actualiza un cliente existente
     */
    public ClienteCatalogo actualizar(ClienteCatalogo cliente) {
        if (cliente == null || cliente.getIdCliente() == null) {
            throw new IllegalArgumentException("Cliente o ID de cliente no puede ser nulo");
        }
        log.info("Actualizando cliente ID: {}, RFC: {}", cliente.getIdCliente(), cliente.getRfc());
        
        // Verificar que el cliente existe y obtener el existente
        Optional<ClienteCatalogo> existenteOpt = clienteCatalogoRepository.findById(cliente.getIdCliente());
        if (existenteOpt.isEmpty()) {
            throw new IllegalArgumentException("Cliente con ID " + cliente.getIdCliente() + " no encontrado");
        }
        
        ClienteCatalogo existente = existenteOpt.get();
        
        // Preservar la fecha de alta del registro original
        if (existente.getFechaAlta() != null) {
            cliente.setFechaAlta(existente.getFechaAlta());
        } else {
            // Si no tenía fecha de alta, establecer la actual
            cliente.setFechaAlta(java.time.LocalDateTime.now());
        }
        
        // Validar campos obligatorios
        if (cliente.getRfc() == null || cliente.getRfc().trim().isEmpty()) {
            throw new IllegalArgumentException("RFC es obligatorio");
        }
        if (cliente.getRazonSocial() == null || cliente.getRazonSocial().trim().isEmpty()) {
            throw new IllegalArgumentException("Razón Social es obligatoria");
        }
        
        try {
            return clienteCatalogoRepository.save(cliente);
        } catch (Exception e) {
            log.error("Error al guardar cliente actualizado", e);
            throw new RuntimeException("Error al actualizar cliente: " + e.getMessage(), e);
        }
    }

    /**
     * Elimina un cliente por ID
     */
    public void eliminar(Long id) {
        if (id == null) {
            throw new IllegalArgumentException("ID de cliente no puede ser nulo");
        }
        log.info("Eliminando cliente ID: {}", id);
        
        // Verificar que el cliente existe
        Optional<ClienteCatalogo> existente = clienteCatalogoRepository.findById(id);
        if (existente.isEmpty()) {
            throw new IllegalArgumentException("Cliente con ID " + id + " no encontrado");
        }
        
        clienteCatalogoRepository.deleteById(id);
        log.info("Cliente ID {} eliminado exitosamente", id);
    }
}

File: facturacion-cibercom/services/clienteCatalogoService.ts
import { apiUrl } from './api';

export interface ClienteDatos {
  idCliente?: number;
  rfc: string;
  razonSocial?: string;
  nombre?: string;
  paterno?: string;
  materno?: string;
  correoElectronico?: string;
  pais?: string;
  domicilioFiscal?: string;
  regimenFiscal?: string;
  registroTributario?: string;
  usoCfdi?: string;
}

export interface ClienteResponse {
  encontrado: boolean;
  cliente?: ClienteDatos | null;
}

class ClienteCatalogoService {
  public buildRfc(iniciales: string, fecha: string, homoclave: string): string {
    return `${(iniciales || '').toUpperCase()}${(fecha || '').toUpperCase()}${(homoclave || '').toUpperCase()}`.trim();
  }

  public async buscarClientePorRFC(rfc: string): Promise<ClienteResponse> {
    const normalized = (rfc || '').toUpperCase();
    if (!normalized) {
      return { encontrado: false, cliente: null };
    }
    try {
      const resp = await fetch(apiUrl(`/catalogo-clientes/${encodeURIComponent(normalized)}`));
      const data = await resp.json().catch(() => null);
      if (!data) {
        return { encontrado: false, cliente: null };
      }
      const cliente: ClienteDatos | undefined = data?.cliente ? {
        rfc: data.cliente.rfc,
        razonSocial: data.cliente.razonSocial,
        nombre: data.cliente.nombre,
        paterno: data.cliente.paterno,
        materno: data.cliente.materno,
        correoElectronico: data.cliente.correoElectronico,
        pais: data.cliente.pais,
        domicilioFiscal: data.cliente.domicilioFiscal,
        regimenFiscal: data.cliente.regimenFiscal,
        registroTributario: data.cliente.registroTributario,
        usoCfdi: data.cliente.usoCfdi,
      } : undefined;
      return { encontrado: Boolean(data?.encontrado && cliente), cliente: cliente ?? null };
    } catch (e) {
      console.error('Error consultando cliente catálogo por RFC', e);
      return { encontrado: false, cliente: null };
    }
  }

  /**
   * Busca clientes por RFC parcial (para autocompletado)
   * Usa el endpoint GET /api/catalogo-clientes/buscar?rfc=...&limit=10
   */
  public async buscarClientesPorRFCParcial(rfc: string, limit: number = 10): Promise<ClienteDatos[]> {
    const normalized = (rfc || '').toUpperCase().trim();
    if (!normalized || normalized.length < 3) {
      return [];
    }
    try {
      const resp = await fetch(apiUrl(`/catalogo-clientes/buscar?rfc=${encodeURIComponent(normalized)}&limit=${limit}`));
      if (resp.ok) {
        const data = await resp.json();
        if (data.clientes && Array.isArray(data.clientes)) {
          return data.clientes.map((c: any) => ({
            rfc: c.rfc || '',
            razonSocial: c.razonSocial,
            nombre: c.nombre,
            paterno: c.paterno,
            materno: c.materno,
            correoElectronico: c.correoElectronico,
            pais: c.pais,
            domicilioFiscal: c.domicilioFiscal,
            regimenFiscal: c.regimenFiscal,
            registroTributario: c.registroTributario,
            usoCfdi: c.usoCfdi,
          }));
        }
      }
      return [];
    } catch (e) {
      console.error('Error buscando clientes por RFC parcial', e);
      return [];
    }
  }

  /**
   * Guarda un nuevo cliente en el catálogo
   * Usa el endpoint POST /api/catalogo-clientes
   */
  public async guardarCliente(cliente: any): Promise<{ success: boolean; idCliente?: number; error?: string }> {
    try {
      const resp = await fetch(apiUrl('/catalogo-clientes'), {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(cliente),
      });
      
      const data = await resp.json();
      
      if (resp.ok && !data.error) {
        return {
          success: true,
          idCliente: data.idCliente,
        };
      } else {
        return {
          success: false,
          error: data.error || 'Error al guardar cliente',
        };
      }
    } catch (e) {
      console.error('Error guardando cliente', e);
      return {
        success: false,
        error: e instanceof Error ? e.message : 'Error desconocido',
      };
    }
  }

  /**
   * Obtiene todos los clientes
   * Usa el endpoint GET /api/catalogo-clientes
   */
  public async obtenerTodos(): Promise<ClienteDatos[]> {
    try {
      const resp = await fetch(apiUrl('/catalogo-clientes'));
      if (resp.ok) {
        const data = await resp.json();
        if (data.clientes && Array.isArray(data.clientes)) {
          return data.clientes.map((c: any) => ({
            idCliente: c.idCliente,
            rfc: c.rfc || '',
            razonSocial: c.razonSocial,
            nombre: c.nombre,
            paterno: c.paterno,
            materno: c.materno,
            correoElectronico: c.correoElectronico,
            pais: c.pais,
            domicilioFiscal: c.domicilioFiscal,
            regimenFiscal: c.regimenFiscal,
            registroTributario: c.registroTributario,
            usoCfdi: c.usoCfdi,
          }));
        }
      }
      return [];
    } catch (e) {
      console.error('Error obteniendo todos los clientes', e);
      return [];
    }
  }

  /**
   * Obtiene clientes paginados
   * Usa el endpoint GET /api/catalogo-clientes?page={page}&size={size}
   */
  public async obtenerTodosPaginados(page: number = 0, size: number = 5): Promise<{
    clientes: ClienteDatos[];
    total: number;
    page: number;
    size: number;
    totalPages: number;
    hasNext: boolean;
    hasPrevious: boolean;
  }> {
    try {
      const resp = await fetch(apiUrl(`/catalogo-clientes?page=${page}&size=${size}`));
      if (resp.ok) {
        const data = await resp.json();
        return {
          clientes: (data.clientes || []).map((c: any) => ({
            idCliente: c.idCliente,
            rfc: c.rfc || '',
            razonSocial: c.razonSocial,
            nombre: c.nombre,
            paterno: c.paterno,
            materno: c.materno,
            correoElectronico: c.correoElectronico,
            pais: c.pais,
            domicilioFiscal: c.domicilioFiscal,
            regimenFiscal: c.regimenFiscal,
            registroTributario: c.registroTributario,
            usoCfdi: c.usoCfdi,
          })),
          total: data.total || 0,
          page: data.page || 0,
          size: data.size || 5,
          totalPages: data.totalPages || 0,
          hasNext: data.hasNext || false,
          hasPrevious: data.hasPrevious || false,
        };
      }
      return {
        clientes: [],
        total: 0,
        page: 0,
        size: 5,
        totalPages: 0,
        hasNext: false,
        hasPrevious: false,
      };
    } catch (e) {
      console.error('Error obteniendo clientes paginados', e);
      return {
        clientes: [],
        total: 0,
        page: 0,
        size: 5,
        totalPages: 0,
        hasNext: false,
        hasPrevious: false,
      };
    }
  }

  /**
   * Obtiene un cliente por ID
   * Usa el endpoint GET /api/catalogo-clientes/id/{id}
   */
  public async obtenerPorId(id: number): Promise<ClienteDatos | null> {
    try {
      const resp = await fetch(apiUrl(`/catalogo-clientes/id/${id}`));
      if (resp.ok) {
        const data = await resp.json();
        if (data.cliente) {
          const c = data.cliente;
          return {
            idCliente: c.idCliente,
            rfc: c.rfc || '',
            razonSocial: c.razonSocial,
            nombre: c.nombre,
            paterno: c.paterno,
            materno: c.materno,
            correoElectronico: c.correoElectronico,
            pais: c.pais,
            domicilioFiscal: c.domicilioFiscal,
            regimenFiscal: c.regimenFiscal,
            registroTributario: c.registroTributario,
            usoCfdi: c.usoCfdi,
          };
        }
      }
      return null;
    } catch (e) {
      console.error('Error obteniendo cliente por ID', e);
      return null;
    }
  }

  /**
   * Actualiza un cliente existente
   * Usa el endpoint PUT /api/catalogo-clientes/{id}
   */
  public async actualizarCliente(id: number, cliente: any): Promise<{ success: boolean; error?: string }> {
    try {
      const resp = await fetch(apiUrl(`/catalogo-clientes/${id}`), {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(cliente),
      });
      
      const data = await resp.json();
      
      if (resp.ok && !data.error) {
        return { success: true };
      } else {
        return {
          success: false,
          error: data.error || 'Error al actualizar cliente',
        };
      }
    } catch (e) {
      console.error('Error actualizando cliente', e);
      return {
        success: false,
        error: e instanceof Error ? e.message : 'Error desconocido',
      };
    }
  }

  /**
   * Elimina un cliente
   * Usa el endpoint DELETE /api/catalogo-clientes/{id}
   */
  public async eliminarCliente(id: number): Promise<{ success: boolean; error?: string }> {
    try {
      const resp = await fetch(apiUrl(`/catalogo-clientes/${id}`), {
        method: 'DELETE',
      });
      
      const data = await resp.json();
      
      if (resp.ok && !data.error) {
        return { success: true };
      } else {
        return {
          success: false,
          error: data.error || 'Error al eliminar cliente',
        };
      }
    } catch (e) {
      console.error('Error eliminando cliente', e);
      return {
        success: false,
        error: e instanceof Error ? e.message : 'Error desconocido',
      };
    }
  }
}

export const clienteCatalogoService = new ClienteCatalogoService();